torch/csrc/jit/tensorexpr/ir_cloner.cpp

```diff
@@ -0,0 +1,343 @@
+    case IRNodeType::kAnd:
+      return new And(lhs_new, rhs_new);
+    case IRNodeType::kOr:
+      return new Or(lhs_new, rhs_new);
+    case IRNodeType::kXor:
+      return new Xor(lhs_new, rhs_new);
+    case IRNodeType::kLshift:
+      return new Lshift(lhs_new, rhs_new);
+    case IRNodeType::kRshift:
+      return new Rshift(lhs_new, rhs_new);
+    default:
+      throw unsupported_dtype();
+  }
+}
+
+Expr* IRCloner::mutate(Add* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(Sub* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(Mul* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(Div* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(Mod* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(And* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(Or* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(Xor* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(Lshift* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(Rshift* v) {
+  return mutate_binary_op(v, this);
+}
+
+Expr* IRCloner::mutate(Max* v) {
+  return mutate_binary_op(v, this, v->propagate_nans());
+}
+
+Expr* IRCloner::mutate(Min* v) {
+  return mutate_binary_op(v, this, v->propagate_nans());
+}
+
+Expr* IRCloner::mutate(CompareSelect* v) {
+  Expr* lhs_new = v->lhs()->accept_mutator(this);
+  Expr* rhs_new = v->rhs()->accept_mutator(this);
+  Expr* retval1_new = v->ret_val1()->accept_mutator(this);
+  Expr* retval2_new = v->ret_val2()->accept_mutator(this);
+  return new CompareSelect(
+      lhs_new,
+      rhs_new,
+      retval1_new,
+      retval2_new,
+      v->compare_select_op(),
+      v->bias());
+}
+
+// NOLINTNEXTLINE
+#define IMM_MUTATE_DEFINE(_1, Name)      \
+  Expr* IRCloner::mutate(Name##Imm* v) { \
+    return v;                            \
+  }
+AT_FORALL_SCALAR_TYPES_AND2(Bool, Half, IMM_MUTATE_DEFINE);
+#undef IMM_MUTATE_DEFINE
+
+Expr* IRCloner::mutate(Cast* v) {
+  Expr* src_value_new = v->src_value()->accept_mutator(this);
+  return new Cast(v->dtype(), src_value_new);
+}
+
+Expr* IRCloner::mutate(BitCast* v) {
+  Expr* src_value_new = v->src_value()->accept_mutator(this);
+  return new BitCast(v->dtype(), src_value_new);
+}
+
+Expr* IRCloner::mutate(Var* v) {
+  return v;
+}
+
+Expr* IRCloner::mutate(Ramp* v) {
+  Expr* base_new = v->base()->accept_mutator(this);
+  Expr* stride_new = v->stride()->accept_mutator(this);
+  return new Ramp(base_new, stride_new, v->lanes());
+}
+
+Expr* IRCloner::mutate(Load* v) {
+  std::vector<Expr*> indices_new;
+  indices_new.reserve(v->indices().size());
+  for (Expr* ind : v->indices()) {
+    indices_new.push_back(ind->accept_mutator(this));
+  }
+  Buf* buf_new = dynamic_cast<Buf*>(v->buf()->accept_mutator(this));
+  return new Load(v->dtype(), buf_new, indices_new);
+}
+
+Expr* IRCloner::mutate(Buf* v) {
```

Comment:
That's a bit awkward that we're cloning Buf, but not cloning Vars. I understand why we might have to do this (since Buf might contain expressions in dims that need to be cloned), but it still could be a surprising behavior. Maybe for consistency we then should be cloning Vars too? Or if that's too breaking, we might consider not cloning Bufs (and its inner expressions). Either way, please document the desired behavior somewhere

Reply:
I debated about this. But went ahead with cloning buffers for the exact reasons you mentioned (because the expressions in dims and initializers). 

Since the difference between cloning Buf and not cloning Var could be confusing as you said, I have reverted to not cloning both of them. The only case I can think of so far which could change the expressions in dims of Buf is buffer compression. And for that case, anyways all the kernel code should be analyzed and it is better to have the clones point to the same Buf. 

If something comes up later, we can revisit this assumption and add cloning for Buf and Var.